tests/push/test_push_rules_http.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2018 New Vector
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List
+from unittest.mock import Mock
+
+from twisted.internet.defer import Deferred
+
+import synapse.rest.admin
+from synapse.logging.context import make_deferred_yieldable
+from synapse.rest.client import login, push_rule, receipts, room
+
+from tests.unittest import HomeserverTestCase
+
+
+class HTTPPusherTests(HomeserverTestCase):
+    servlets = [
+        synapse.rest.admin.register_servlets_for_client_rest_resource,
+        room.register_servlets,
+        login.register_servlets,
+        receipts.register_servlets,
+        push_rule.register_servlets,
+    ]
+    user_id = True
+    hijack_auth = False
+
+    def default_config(self):
+        config = super().default_config()
+        config["start_pushers"] = True
+        return config
+
+    def make_homeserver(self, reactor, clock):
+        self.push_attempts: List[tuple[Deferred, str, dict]] = []
+
+        m = Mock()
+
+        def post_json_get_json(url, body):
+            d: Deferred = Deferred()
+            self.push_attempts.append((d, url, body))
+            return make_deferred_yieldable(d)
+
+        m.post_json_get_json = post_json_get_json
+
+        hs = self.setup_test_homeserver(proxied_blacklisted_http_client=m)
+
+        return hs
+
+    def _make_user_with_pusher(self, username):
```

Comment:
We've been trying to slow get type hints everywhere, especially in easy spots:

```suggestion
    def _make_user_with_pusher(self, username: str) -> Tuple[str, str]:
    	"""
    	Register a user, configure an HTTP pusher for them.
    	
    	Return:
    		A tuple of the user name & access token.
    	"""
```

Reply:
https://github.com/matrix-org/synapse/pull/12188/commits/e10edf91bfc1da7b29d6cdc90d874d8fb24ab96d